### PR TITLE
feat(linter): implement `no-misused-new`

### PR DESCRIPTION
**Pull request by cijiugechu**
related issue: #503

### CODE REVIEW

File: crates/oxc_linter/src/snapshots/no_misused_new.snap [hunk 1]
```diff
@@ -0,0 +1,47 @@
+---
+source: crates/oxc_linter/src/tester.rs
+expression: no_misused_new
+---
+  ⚠ typescript-eslint(no-misused-new): Interfaces cannot be constructed, only classes.
+   ╭─[no_misused_new.tsx:1:1]
+ 1 │ interface I { new (): I; constructor(): void;}
```
Comment (Boshen) [inline crates/oxc_linter/src/snapshots/no_misused_new.snap:7]: nitpick: labeling the `new` (method name) is sufficient I think

File: crates/oxc_linter/src/rules/typescript/no_misused_new.rs [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+#[error("typescript-eslint(no-misused-new): Interfaces cannot be constructed, only classes.")]
+#[diagnostic(
+    severity(warning)
+)]
+struct NoMisusedNewInterfaceDiagnostic(#[label] pub Span);
+
+#[derive(Debug, Error, Diagnostic)]
+#[error("typescript-eslint(no-misused-new): Class cannot have method named `new`.")]
+#[diagnostic(
+    severity(warning)
```
Comment (Boshen) [inline crates/oxc_linter/src/rules/typescript/no_misused_new.rs:21]: nitpick: do we have any help message for why `new` and `constructor` is bad?

**Review by Boshen — CHANGES_REQUESTED** [review]

File: crates/oxc_linter/src/rules/typescript/no_misused_new.rs (hunk 2 above)
Comment (cijiugechu) [reply crates/oxc_linter/src/rules/typescript/no_misused_new.rs]: There are no corresponding help messages in the original typescript lint rules, so I'm not sure if this help message is appropriate.

File: crates/oxc_linter/src/rules/typescript/no_misused_new.rs [hunk 3]
```diff
@@ -4,21 +4,23 @@ use oxc_diagnostics::{
...
+    help("Consider removing this method from your interface.")
 )]
 struct NoMisusedNewInterfaceDiagnostic(#[label] pub Span);
 
 #[derive(Debug, Error, Diagnostic)]
 #[error("typescript-eslint(no-misused-new): Class cannot have method named `new`.")]
 #[diagnostic(
-    severity(warning)
+    severity(warning),
+    help("This method name is confusing , consider renaming the method to `constructor`")
```
Comment (Boshen) [inline crates/oxc_linter/src/rules/typescript/no_misused_new.rs:23]: there's an extra space in frontend `,`

File: crates/oxc_linter/src/rules/typescript/no_misused_new.rs [hunk 4]
```diff
@@ -68,15 +70,17 @@ impl Rule for NoMisusedNew {
                  let TSTypeName::IdentifierName(id) = &type_ref.type_name &&
                  id.name == decl_name {
 
-                     ctx.diagnostic(NoMisusedNewInterfaceDiagnostic(sig.span));
+                     ctx.diagnostic(NoMisusedNewInterfaceDiagnostic(
+                      Span::new(sig.span.start, sig.span.start + 3)
```
Comment (Boshen) [inline crates/oxc_linter/src/rules/typescript/no_misused_new.rs:74]: Should this be span of the `id`?

**Review by Boshen — CHANGES_REQUESTED** [review]

File: crates/oxc_linter/src/rules/typescript/no_misused_new.rs (hunk 4 above)
Comment (Boshen) [reply crates/oxc_linter/src/rules/typescript/no_misused_new.rs:74]: By the way, where is the comparison to `new`?

Comment (cijiugechu) [reply crates/oxc_linter/src/rules/typescript/no_misused_new.rs:74]: > Should this be span of the `id`?

The `id` here refers to the return type, not the method name.

> By the way, where is the comparison to new or constructor? TSConstructSignatureDeclaration?

The two are distinguished by the `TSConstructSignatureDeclaration` and the `TSMethodSignature`.

**Review by Boshen — APPROVED** [review]

**Pull request merged**